### ISSUE TITLE
Intermittent/concurrency error on serialization in aspnetcore odata 7.5.4

### ISSUE THREAD

**Issue by houbi56** [OData/WebApi#2390]
Getting intermittent serialization errors in 7.5.4 under load, worked without issues in 7.5.2

We use a class IgnoreDefaultEntityPropertiesSerializer : ODataResourceSerializer that on CreateStructuralProperty returns null if 
the value of the created ODataproperty is of primitive type and has a default value. 
Prop is created by calling base.CreateStructuralProperty(structuralProperty, resourceContext);

The error happens when multiple (16) concurrent calls are made from a webpage, typically 10-20% of the time one of the calls will fail with the error:
System.InvalidOperationException: The EDM instance of type '[xxx Nullable=True]' is missing the property 'yyy'.. 

--
System.InvalidOperationException: The EDM instance of type '[xxx Nullable=True]' is missing the property 'yyy'.
at Microsoft.AspNet.OData.ResourceContext.GetPropertyValue(String propertyName)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralProperty(IEdmStructuralProperty structuralProperty, ResourceContext resourceContext)
   at [ournamespace].Server.Configuration.Serialization.IgnoreDefaultEntityPropertiesSerializer.CreateStructuralProperty(IEdmStructuralProperty structuralProperty, ResourceContext resourceContext) 
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralPropertyBag(SelectExpandNode selectExpandNode, ResourceContext resourceContext)

---

**Comment by houbi56** [comment]
Posted in wrong repo, closing.

**Comment by xuzhg** [comment]
@houbi56 Could you please share us a simple repro to dig more?

**Comment by houbi56** [comment]
https://github.com/houbi56/OData754ConcurrencyError

**Comment by houbi56** [comment]
See github repo for repro
I get the exception below when performing 20 concurrent calls. Seems that $select as of 7.5.4 is no longer thread-safe?

```
var c = new HttpClient();
var url = "http://localhost:5000/odata/ODataEntity?$select=Id,SomeOtherProperty";
Parallel.ForEach(Enumerable.Range(1, 20), async (i) =>
{
    var r = await c.GetAsync(url);
});
```

--

System.InvalidOperationException: The EDM instance of type '[OData_754_Error.ODataEntity Nullable=True]' is missing the property 'SomeProperty'.
   at Microsoft.AspNet.OData.ResourceContext.GetPropertyValue(String propertyName)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralProperty(IEdmStructuralProperty structuralProperty, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateStructuralPropertyBag(SelectExpandNode selectExpandNode, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.CreateResource(SelectExpandNode selectExpandNode, ResourceContext resourceContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSerializer.WriteResourceAsync(Object graph, ODataWriter writer, ODataSerializerContext writeContext, IEdmTypeReference expectedType)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteResourceSetAsync(IEnumerable enumerable, IEdmTypeReference resourceSetType, ODataWriter writer, ODataSerializerContext writeContext)
   at Microsoft.AspNet.OData.Formatter.Serialization.ODataResourceSetSerializer.WriteObjectAsync(Object graph, Type type, ODataMessageWriter messageWriter, ODataSerializerContext writeContext)
   at Microsoft.AspNet.OData.Formatter.ODataOutputFormatterHelper.WriteToStreamAsync(Type type, Object value, IEdmModel model, ODataVersion version, Uri baseAddress, MediaTypeHeaderValue contentType, IWebApiUrlHelper internaUrlHelper, IWebApiRequestMessage internalRequest, IWebApiHeaders internalRequestHeaders, Func`2 getODataMessageWrapper, Func`2 getEdmTypeSerializer, Func`2 getODataPayloadSerializer, Func`1 getODataSerializerContext)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ObjectResultExecutor.ExecuteAsyncEnumerable(ActionContext context, ObjectResult result, Object asyncEnumerable, Func`2 reader)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResultFilterAsync>g__Awaited|29_0[TFilter,TFilterAsync](ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResultExecutedContextSealed context)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.ResultNext[TFilter,TFilterAsync](State& next, Scope& scope, Object& state, Boolean& isCompleted)
   at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeResultFilters()
......

**Comment by houbi56** [comment]
Did some more digging. 
This PR breaks it [https://github.com/OData/WebApi/pull/2362](https://github.com/OData/WebApi/pull/2362).
Fields are used in an actionfilter which (according to MS Docs) is not threadsafe.

EnableQueryAttribute


        private ODataQueryOptions CreateAndValidateQueryOptions(HttpRequest request, ODataQueryContext queryContext)
        {
            if (_queryValidationRunBeforeActionExecution)
            {
                //**Below is not thread safe and returns data belonging to different http request** 
                return _processedQueryOptions  context;
            }

            ODataQueryOptions queryOptions = new ODataQueryOptions(queryContext, request);

            ValidateQuery(request, queryOptions);

            return queryOptions;
        }

**Comment by houbi56** [comment]
Confirm fixed in #2397
